File: src/main/java/com/marklynch/jsonclient/io/Worker.java
package com.marklynch.jsonclient.io;

import com.marklynch.jsonclient.KotlinWorker;
import com.marklynch.jsonclient.io.api.Server;
import com.marklynch.jsonclient.io.api.response.ResponseObject;
import org.jetbrains.annotations.NotNull;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class Worker {

    private final Server server = new Server();

    public void go() {
        server.callQuestionRequest(questionCallback);
    }

    private Callback<ResponseObject> questionCallback = new Callback<ResponseObject>() {
        @Override
        public void onResponse(@NotNull Call<ResponseObject> call, Response<ResponseObject> response) {
            if (!response.isSuccessful() || response.body() == null) {
                //ERROR
                System.out.println("testQuestionCallback.onResponse ERROR");
                System.out.println("call.request().url() = " + call.request().url());
                System.out.println("response.code() = " + response.code());
                System.out.println("response.body() = " + response.body());
            } else {
                //SUCCESS
                System.out.println("testQuestionCallback.onResponse SUCCESS");
                System.out.println("call.request().url() = " + call.request().url());
                System.out.println("response.code() = " + response.code());
                System.out.println("response.body() = " + response.body());
                String answer = doWork(response.body());
                server.callAnswerGetRequest(answer, testAnswerCallback);
                server.callAnswerPostRequest(answer, testAnswerCallback);
            }
        }

        @Override
        public void onFailure(@NotNull Call<ResponseObject> call, @NotNull Throwable t) {
            //FAILURE
            System.out.println("testQuestionCallback.onFailure ERROR");
            System.out.println("call.request().url() = " + call.request().url());
            t.printStackTrace();
        }
    };

    private Callback<Void> testAnswerCallback = new Callback<Void>() {
        @Override
        public void onResponse(@NotNull Call<Void> call, Response<Void> response) {
            if (!response.isSuccessful()) {
                //ERROR
                System.out.println("testAnswerCallback.onResponse ERROR");
                System.out.println("call.request().url() = " + call.request().url());
                System.out.println("call.request().body() = " + call.request().body());
                System.out.println("response.code() = " + response.code());
                System.out.println("response.body() = " + response.body());
            } else {
                //SUCCESS
                System.out.println("testAnswerCallback.onResponse SUCCESS");
                System.out.println("call.request().url() = " + call.request().url());
                System.out.println("call.request().body() = " + call.request().body());
                System.out.println("response.code() = " + response.code());
                System.out.println("response.body() = " + response.body());
            }
        }

        @Override
        public void onFailure(@NotNull Call<Void> call, @NotNull Throwable t) {
            //FAILURE
            System.out.println("testAnswerCallback.onFailure ERROR");
            System.out.println("call.request().url() = " + call.request().url());
            t.printStackTrace();
        }
    };

    private String doWork(ResponseObject response) {

        String answer = KotlinWorker.Companion.doWork(response);

        System.out.println("doWork answer = " + answer);
        return answer;
    }

}

File: src/main/kotlin/com/marklynch/jsonclient/io/Worker.kt
package com.marklynch.jsonclient.io

import com.marklynch.jsonclient.io.api.Server
import com.marklynch.jsonclient.io.api.response.ResponseObject
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class Worker {

    private val server = Server()

    private val questionCallback = object : Callback<ResponseObject> {
        override fun onResponse(call: Call<ResponseObject>, response: Response<ResponseObject>) {
            if (!response.isSuccessful || response.body() == null) {
                //ERROR
                println("testQuestionCallback.onResponse ERROR")
                println("call.request().url() = " + call.request().url)
                println("response.code() = " + response.code())
            } else {
                //SUCCESS
                println("testQuestionCallback.onResponse SUCCESS")
                println("call.request().url() = " + call.request().url)
                println("response.code() = " + response.code())
                println("response.body() = " + response.body())
                val answer = doWork(response.body())
                server.callAnswerGetRequest(answer, testAnswerCallback)
                server.callAnswerPostRequest(answer, testAnswerCallback)
            }
        }

        override fun onFailure(call: Call<ResponseObject>, t: Throwable) {
            //FAILURE
            println("testQuestionCallback.onFailure ERROR")
            println("call.request().url() = " + call.request().url)
            t.printStackTrace()
        }
    }

    private val testAnswerCallback = object : Callback<Void> {
        override fun onResponse(call: Call<Void>, response: Response<Void>) {
            if (!response.isSuccessful) {
                //ERROR
                println("testAnswerCallback.onResponse ERROR")
                println("call.request().url() = " + call.request().url)
                println("call.request().body() = " + call.request().body!!)
                println("response.code() = " + response.code())
                println("response.body() = " + response.body())
            } else {
                //SUCCESS
                println("testAnswerCallback.onResponse SUCCESS")
                println("call.request().url() = " + call.request().url)
                println("call.request().body() = " + call.request().body!!)
                println("response.code() = " + response.code())
                println("response.body() = " + response.body()!!)
            }
        }

        override fun onFailure(call: Call<Void>, t: Throwable) {
            //FAILURE
            println("testAnswerCallback.onFailure ERROR")
            println("call.request().url() = " + call.request().url)
            t.printStackTrace()
        }
    }

    fun go() {
        server.callQuestionRequest(questionCallback)
    }

    private fun doWork(response: ResponseObject?): String {
        val answer = response.toString()
        println("doWork answer = $answer")
        return answer
    }

}
